server/api/src/rsDataObjUnlink.cpp

```diff
@@ -547,125 +564,95 @@ int dataObjUnlinkS(rsComm_t* rsComm,
-                std::string location;
-                irods::error ret = irods::get_loc_for_hier_string( dataObjInfo->rescHier, location );
-                if ( !ret.ok() ) {
-                    irods::log( PASSMSG( "dataObjUnlinkS - failed in get_loc_for_hier_string", ret ) );
-                    return ret.code();
-                }
+            return remote_flag;
+        }
 
-                rodsHostAddr_t addr;
-                rodsServerHost_t *rodsServerHost = 0;
+        // If the vault path is validated (ret.ok() is true), that indicates that the replica
+        // resides in an iRODS resource vault. If the user is not a rodsadmin, unregistering
+        // a replica from inside an iRODS resource vault is not allowed.
+        std::string out_path;
+        if (const auto ret = resc_mgr.validate_vault_path(info.physical_path().data(), host, out_path); ret.ok()) {
+            irods::log(LOG_ERROR, fmt::format(
+                "[{}:{}] - non-admin users cannot unregister a replica in a vault [path=[{}], hierarchy=[{}]]",
+                __FUNCTION__, __LINE__, info.logical_path(), info.hierarchy()));
 
-                memset( &addr, 0, sizeof( addr ) );
-                rstrcpy( addr.hostAddr, location.c_str(), NAME_LEN );
-                int remoteFlag = resolveHost( &addr, &rodsServerHost );
-                if ( remoteFlag < 0 ) {
-                    // error condition?
-                }
+            return CANT_UNREG_IN_VAULT_FILE;
+        }
+    }
 
-                /* unregistering but not an admin user */
-                std::string out_path;
-                ret = resc_mgr.validate_vault_path( dataObjInfo->filePath, rodsServerHost, out_path );
-                if ( ret.ok() ) {
-                    /* in the vault */
-                    std::stringstream msg;
-                    msg << "unregistering a data object which is in a vault [";
-                    msg << dataObjInfo->filePath << "]";
-                    irods::log( PASSMSG( msg.str(), ret ) );
-                    return CANT_UNREG_IN_VAULT_FILE;
-                }
+    // The UNREG_OPR type is used to indicate that the replica should not be physically unlinked,
+    // but only unregistered. Therefore, only attempt physical unlink if this is not an UNREG_OPR.
+    if (UNREG_OPR != dataObjUnlinkInp->oprType) {
+        const auto cond_input = irods::experimental::make_key_value_proxy(dataObjUnlinkInp->condInput);
+        info.in_pdmo(cond_input.contains(IN_PDMO_KW) ? cond_input.at(IN_PDMO_KW).value() : "");
+
+        if (const auto ec = l3Unlink(rsComm, info.get()); ec < 0) {
+            irods::log(LOG_NOTICE, fmt::format(
+                "[{}:{}] - failed to physically unlink replica "
+                "[error_code=[{}], path=[{}], hierarchy=[{}]",
+                __FUNCTION__, __LINE__, ec, info.logical_path(), info.hierarchy()));
+
+            // TODO: why do we allow NOENT and EACCES to unregister?
```

Comment:
b/c the physical file already has a problem?   NOENT means it's not even there... and EACCES means the service account can't read it anyway...

Reply:
I think I will revisit this on the pass to prevent unlinks on locked data objects.